Add support for Windows NPM

MetaMmodern opened the pull request:
# Description

These changes add support for Windows OS. Was tested on Windows 10 (x64), server and web-client start now.

> :warning: **I am not a haskell developer!**: My code may be ugly and even disguisting, but that's what I've achived by googling)

### Code explanation

1. The reason why npm was not starting is because Windows Process API appends .exe automatically to the executable filename, however npm as program does not provide any .exe file, only .cmd and .ps1. (explained here: https://stackoverflow.com/questions/43139364/createprocess-weird-behavior-with-files-without-extension) The first thing to do is to support .cmd for windows only. That's implemented in `oSSpecificNpm` variable. _(Screenshot of ghci session for proof)_
![image](https://user-images.githubusercontent.com/42899786/176509688-7463a0d8-0e11-4419-a092-cae048aa835d.png)

2. When we run `npm.cmd` from inside haskell it will complain that it can't find the `npm-cli`. The reason and solution is explained here: https://stackoverflow.com/a/44820337 . The implementation is made in `compileOsSpecificNodeCommand` Common function and in `src/Wasp/Generator/Job/Process.hs` file. _(Screenshot as demo)_
![image](https://user-images.githubusercontent.com/42899786/176509815-37d486fd-97ad-4bc5-b34e-885001185d17.png)


### Things to do further

Do NOT update docs and do not mention Windows support, since there are still following issues persistent(at least in my env):
1. Some path's in generated js files are **incorrect** for windows (backslashes do not work).
![image](https://user-images.githubusercontent.com/42899786/176510750-65d456ca-3189-4183-90b6-2abe4665d9fc.png)

2. Some issues with db and/or with encoding during db initialization/migration.
![image](https://user-images.githubusercontent.com/42899786/176510770-3e9e3cbe-98dd-4391-b153-36bddd9c9782.png)

3. UI did not render for me😢.
![image](https://user-images.githubusercontent.com/42899786/176511303-fc883027-3a83-49ef-a410-92b184bad2e2.png)


Fixes #48

## Type of change

Please select the option(s) that is more relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] This change requires a documentation update

MetaMmodern committed:
add support for windows NPM

MetaMmodern committed:
lint: remove redundant import

Martinsos reviewed (changes requested):
@MetaMmodern Super awesome that we are making progress in the Windows direction!!

You did a great job taking into account that you don't know Haskell :). I left some comments, mostly it comes down to figuring out what exactly we want to do. My previous bad naming makes things a bit more complex but I think we can fix that now also.

Martinsos commented:
Ah I realize now you explained in the description why we are doing the `cmd.exe` thing! Ok I will check it out sorry about that -> I have to run now but I should be able to get back to it in in an hour or two, I will read it then and write more educated comment!

MetaMmodern commented on waspc/src/Wasp/Generator/Common.hs:
```diff
@@ -38,3 +42,12 @@ npmVersionRange =
 prismaVersion :: SV.Version
 prismaVersion = SV.Version 3 15 2
+
+oSSpecificNpm :: [Char]
```
@Martinsos for Strings--got it, did not think haskell has those) For exmplanation comment--no problem. Should I leave a text explanation or leave a link to SO?

Martinsos replied:
Maybe best to do both: text explanation + SO link!

Martinsos commented:
Maybe both is best!

On Thu, 30 Jun 2022, 11:08 Bunyamin Shabanov, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In waspc/src/Wasp/Generator/Common.hs
> <https://github.com/wasp-lang/wasp/pull/658#discussion_r910791614>:
>
> > @@ -38,3 +42,12 @@ npmVersionRange =
>
>  prismaVersion :: SV.Version
>  prismaVersion = SV.Version 3 15 2
> +
> +oSSpecificNpm :: [Char]
>
> @Martinsos <https://github.com/Martinsos> for Strings--got it, did not
> think haskell has those) For exmplanation comment--no problem. Should I
> leave a text explanation or leave a link to SO?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wasp-lang/wasp/pull/658#discussion_r910791614>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALXFBZ5YMBSJY5ZTTKEK7TVRVPZJANCNFSM52GXRSUQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Martinsos commented on waspc/src/Wasp/Generator/Job/Process.hs:
```diff
@@ -102,7 +102,8 @@ runNodeCommandAsJob fromDir command args jobType chan = do
       if SV.isVersionInRange nodeVersion C.nodeVersionRange
         then do
-          let process = (P.proc command args) {P.cwd = Just $ SP.fromAbsDir fromDir}
+          let (specificCommand, specificArgs) = C.compileOsSpecificNodeCommand command args
```
Ok, so now I read what you wrote about `cmd.exe` and read the SO answer you linked to!
So interesting, it seems that `npm.cmd` script is written in such way that calling it programmatically can make it work wrong, so by adding `cmd.exe /c` in front we ensure it works correctly.

The question is then: Do we want to use this trick only when calling `npm`, or do we want to use it when calling any commands on Windows? I am thinking maybe it is enough to use it just for calling `npm`, at least for now?

In that case, let's use it like that, only for `npm`. So what we could do is rename this function to
```hs
buildNpmCmdWithArgs :: [String] -> (String, [String])
```
and that is it? And then we call it only in places where we need to call `npm`?

Martinsos commented:
Alright, read about cmd.exe and left the comment, that is it from me for this first iteration of review!

Martinsos commented:
Oh btw we use ormolu for formatting Haskell code -> is that something you could easily set up, so that we can get the CI working? If not, don't worry, I can take care of that for you later.

Martinsos committed:
Fixed path in generated JS file that was not always Posix.

MetaMmodern commented:
> Oh btw we use ormolu for formatting Haskell code -> is that something you could easily set up, so that we can get the CI working? If not, don't worry, I can take care of that for you later.

Um, I'll try but not now, later evening)

Martinsos committed:
Now we decode output of subprocesses with locale encoding instead of utf8.

Martinsos commented:
In the meantime I made commits that should fix the two problems that you identified @MetaMmodern (wrong path and problem with \xad) -> I pushed them onto your `main` branch, so make sure to pull it before you continue.
This should enable us to move forward again :)! I believe there will be another problem, which is the fact that `script` command does not exist on Windows, and we currently need it for smth, but let's see if that is a problem or not first and then we can consider it more.

MetaMmodern committed:
fix: types to be Strings

MetaMmodern committed:
lint: clear the namespace

MetaMmodern commented on waspc/src/Wasp/Generator/Common.hs:
```diff
@@ -3,10 +3,14 @@ module Wasp.Generator.Common
 
 import qualified Wasp.SemanticVersion as SV
+import System.Info (os)
+import Data.List
```
Done

MetaMmodern committed:
fix: bad naming

MetaMmodern committed:
add function comments

MetaMmodern committed:
format: with ormulu

MetaMmodern commented:
@Martinsos for db migration similar error: 
![image](https://user-images.githubusercontent.com/42899786/177006691-68f04ea5-d607-498b-bc38-328177c9dcf3.png)

Martinsos commented on waspc/src/Wasp/Generator/Common.hs:
```diff
@@ -38,3 +42,16 @@ npmVersionRange =
 prismaVersion :: SV.Version
 prismaVersion = SV.Version 3 15 2
+
+-- | Adds .cmd extension to callable npm command if ran on Windows, since npm does note have .exe on Windows.
+--   The reason explained here: https://stackoverflow.com/questions/43139364/createprocess-weird-behavior-with-files-without-extension
+oSSpecificNpm :: String
+oSSpecificNpm = "npm" ++ if os /= "mingw32" then "" else ".cmd"
```
I usually try to document as much as I can with the code, but if I can't capture it in the code, then it goes in the comments, with as little duplication as possible.

Also, I differ between comments explaining functions API (what code calling it needs to know about it) and comments explaining implementation details (what developer maintaining the innards of the function needs to know about it).

So in this case, I would do some modifications like this:
```suggestion
npmCmd :: String
npmCmd =
  -- Windows add ".exe" to command, when calling it programmatically, if it doesn't
  -- have an extension already, meaning that calling `npm` actually calls `npm.exe`.
  -- However, there is no `npm.exe` on Windows, instead there is `npm` or `npm.cmd`, so we make sure here to call `npm.cmd`.
  -- Extra info: https://stackoverflow.com/questions/43139364/createprocess-weird-behavior-with-files-without-extension .
  "npm" ++ if os /= "mingw32" then "" else ".cmd"
```

Martinsos commented on waspc/src/Wasp/Generator/Common.hs:
```diff
@@ -38,3 +42,16 @@ npmVersionRange =
+
+-- | Changes an npm command to a cmd.exe command on Windows only. Calling npm from API causes troubles.
+--   The reason and solution exaplined here: https://stackoverflow.com/a/44820337 
+buildNpmCmdWithArgs :: String -> [String] -> (String, [String])
```
We said this is to be used only for calling `npm`. But in that case, we don't need `command` as input, right? Since we know it is `npm`?

MetaMmodern replied:
fixed

Martinsos commented on waspc/src/Wasp/Generator/Common.hs:
```diff
@@ -38,3 +42,16 @@ npmVersionRange =
+oSSpecificNpm :: String
+oSSpecificNpm = "npm" ++ if os /= "mingw32" then "" else ".cmd"
+
+-- | Changes an npm command to a cmd.exe command on Windows only. Calling npm from API causes troubles.
```
Similar like above, this is addressing implementation details. I would make sure this comment talks about what function does from the outside, now how it does it, and details can go inside of function. We might not even need this comment if it is clean what it does from the function name + signature.

MetaMmodern replied:
moved comments into implementation

Martinsos commented on waspc/src/Wasp/Generator/Job/Process.hs:
```diff
@@ -94,15 +91,30 @@ runProcessAsJob process jobType chan =
       if SV.isVersionInRange nodeVersion C.nodeVersionRange
         then do
-          let process = (P.proc command args) {P.cwd = Just $ SP.fromAbsDir fromDir}
+          let (specificCommand, specificArgs) = C.buildNpmCmdWithArgs command args
```
Now that we renamed the function to `buildNpmCmdWithArgs`, and that it is used only for running `npm` + some args, I don't think we want to call it here anymore, since here we are dealing with any command that requires node installed, which can also be other stuff that is not npm, for example `npx`.

MetaMmodern replied:
Done)

Martinsos commented:
Awesome @MetaMmodern !

I left couple more comments, some are just comments/naming, one is about where to call the "buildNpmCmdWithArgs" function.

As for encoding issue with prisma db migrate -> that happens now due to us not supporting the encoding that you use in your terminal, which is CP866 (russian cyrilic).
I found this library now https://hackage.haskell.org/package/encoding-0.8.7/docs/Data-Encoding.html, once we add that we should have support for CP866 and many other encodings -> I just need to find some time to do it properly since it is not super simple due to them complicating things a bit with Monads!

Ha yes it takes long time for cabal since there is a lot to compile really, Haskell is more complex language than JS, but usually it is so slow only on the first build!

MetaMmodern committed:
Update waspc/src/Wasp/Generator/Common.hs

applied suggested change

Co-authored-by: Martin Šošić <Martinsos@users.noreply.github.com>

MetaMmodern committed:
fix: change call signature to npm only

MetaMmodern committed:
Merge branch 'main' of https://github.com/MetaMmodern/wasp

MetaMmodern committed:
fix: comments placement

Martinsos committed:
Fixed compiler errors, refactored node job function a bit.

Martinsos committed:
Renamed node-cmd-as-job function.

Martinsos committed:
Added additional encodings support to Wasp.

Martinsos reviewed (approved):
@MetaMmodern this should be it! In order to speed the merging up a bit, I fixed one tiny compiler error that somehow sneaked in, did a tiny bit of left-over refactoring, and I added more encodings, that should resolve the issue you had with CP861!

Would you like to give it a finally try, to see if the issue with CP861 is now resolved?

MetaMmodern commented:
> @MetaMmodern this should be it! In order to speed the merging up a bit, I fixed one tiny compiler error that somehow sneaked in, did a tiny bit of left-over refactoring, and I added more encodings, that should resolve the issue you had with CP861!
> 
> Would you like to give it a finally try, to see if the issue with CP861 is now resolved?

in an hour or so) lil bit later

MetaMmodern commented:
@Martinsos what's with build?

Martinsos commented:
@MetaMmodern hm good question, it doesn't seem to show any error messages -> I restarted it, let's see if that helps!

Martinsos committed:
Modified encodings flag.

Martinsos commented:
@MetaMmodern I think I found the reason -> there was this flag for this new `encoding` package that sometimes causes problems. I turned it off since we don't use it -> let's see now if it helps.
What sucks is that `encodings` package takes really long time to build, 5-10 minutes, I am not sure why, it is reported on their GH repo as an issue but not resolved yet. But there aren't really any better alternatives out there.

TODO for myself: maybe try with the `knob` package -> wrap bytestrings into file handlers and let withUtf8 handle it. But does withUtf8 support wide array of encodings or not?

MetaMmodern commented:
@Martinsos maybe there is a way to apply caching to library builds? Like separate those into some kind of .bin files and use cache if dependencies did not change)

Martinsos commented:
It builds now @MetaMmodern

MetaMmodern commented:
@Martinsos nnnnnope)
![image](https://user-images.githubusercontent.com/42899786/177189747-06a070ad-33b4-41e1-90c7-192101fb550b.png)

Martinsos commented:
Wooo great! This was unfortunately expected, but we took care of encoding
yay! This is furthest we got right?

On Mon, 4 Jul 2022, 18:06 Bunyamin Shabanov, ***@***.***>
wrote:

> @Martinsos <https://github.com/Martinsos> nnnnnope)
> [image: image]
> <https://user-images.githubusercontent.com/42899786/177189747-06a070ad-33b4-41e1-90c7-192101fb550b.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wasp-lang/wasp/pull/658#issuecomment-1173969937>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALXFB32RHA74DFD325MC33VSMDWXANCNFSM52GXRSUQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

MetaMmodern commented:
> Wooo great! This was unfortunately expected, but we took care of encoding
> yay! This is furthest we got right?
> 
> On Mon, 4 Jul 2022, 18:06 Bunyamin Shabanov, ***@***.***>
> wrote:
> 
> > @Martinsos <https://github.com/Martinsos> nnnnnope)
> > [image: image]
> > <https://user-images.githubusercontent.com/42899786/177189747-06a070ad-33b4-41e1-90c7-192101fb550b.png>
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <https://github.com/wasp-lang/wasp/pull/658#issuecomment-1173969937>, or
> > unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/AALXFB32RHA74DFD325MC33VSMDWXANCNFSM52GXRSUQ>
> > .
> > You are receiving this because you were mentioned.Message ID:
> > ***@***.***>
> >
> 

Well, as I understood--yes.

Martinsos committed:
Fixed more for Windows

Martinsos commented:
I managed to set up my Windows to run Haskell, so now I can also test!

Ok, so I did following, as a quick hack in the place where we now use `script`and it seems to be enough to get prisma db migrate-dev to work (at least for now):
```
let job = runNodeDependentCommandAsJob J.Db serverDir ("cmd.exe", pure $ unwords $ "/c" : npxPrismaMigrateCmd)
```
I will try to properly add this to the code.

Regarding your encoding @MetaMmodern -> is there a way I can also set my encoding to be like yours, so I can test? I tried setting language to Croatian, and to Russian, but it uses CP437 in that case and I don't have any issues with encodings, even if I just do decodeUTF8 -> ok, escape characters are often not correctly shown, but it doesn't break. So how can I reproduce it to have the same situation as you? Because I am not happy with the current encoding solution I made, it works but it has its problems.

MetaMmodern commented:
@Martinsos hey, congrats on setting up Windows. For encoding: try running `chcp 866` in cmd window. Should work. Got it from here:  https://superuser.com/questions/1170656/windows-10-terminal-encoding

Martinsos commented:
Thanks, I tried that @MetaMmodern ! But I can't reproduce the failure to parse, the one with `\xad` error. 

I did a final commit, which:
1) Removes usage of script for Windows -> for me `wasp db migrate-dev` now works! But it didn't ask me any questions. I know sometimes it asks, and I wonder if in that case it will not work. But I don't know how to get it to ask. Maybe you can try running `wasp db migrate-dev`, see how it goes?
2) I reversed encoding solution to the most simple one, which works for me regardless of encoding. I wonder if this will still work for you, or will cause problems again.

TLDR; Can you give it one last spin? I did as much as I could on my Windows, for me it works now, but I wonder if it will work for you with your CP861 encoding (worked for me when I used chcp 861 but it might be different).

Martinsos committed:
Finally fixed encodings?

MetaMmodern commented:
@Martinsos What does this mean? 
![image](https://user-images.githubusercontent.com/42899786/177327181-05b5ff18-a0b1-429f-96e1-8e4cd699b059.png)
Do I need to have Postgres Installed to run WASP? Do I need to manually add this env variable or is it set by WASP CLI?

Martinsos commented:
Oho awesome you got this far!!! So it all works!

Yes it seems you need postgre database and provide its connection URL via
this variable. Which app are you running? In docs you have detailed
instructions on how to run postgre dB, includin one liner to get it going
quickly via docker.

On Tue, 5 Jul 2022, 14:27 Bunyamin Shabanov, ***@***.***>
wrote:

> @Martinsos <https://github.com/Martinsos> What does this mean?
> [image: image]
> <https://user-images.githubusercontent.com/42899786/177327181-05b5ff18-a0b1-429f-96e1-8e4cd699b059.png>
> Do I need to have Postgres Installed to run WASP? Do I need to manually
> add this env variable or is it set by WASP CLI?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wasp-lang/wasp/pull/658#issuecomment-1175001065>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALXFB22MAPAXESN5ENKIPLVSQS3HANCNFSM52GXRSUQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

MetaMmodern commented:
@Martinsos okay, I'll try it now.

Martinsos commented:
Awesome @MetaMmodern well peremoga it is then :D, I can't believe we have Windows support working!! Thanks again, without you pushing this, we would not have it working soon for sure.
Let me keep this PR open few more days so I can do some final tweaks regarding the encodings + rebase it on newest master since some changes are coming there that we will need to adapt to, but everything important is done! Yay :)

Martinsos committed:
Implemented encoding function for ByteString.

Martinsos committed:
Removed knob library and now using direct decoding.

Martinsos commented:
@MetaMmodern , when you catch a moment, could you run one last test? I now made a change in the code we use for decoding, so it is less convoluted + doesn't need external library -> in theory it should be the same, but if you can give it a try (wasp db migrate-dev) I would feel  sure it works. Unfortunately I still haven't managed to replicate your situation with encoding in order to test the situation you had.

Btw is your output messy in any way? For me, on Windows, it is often not colored when it should be, plus some chars get replaced with `?`, the exotic ones.

MetaMmodern commented:
@Martinsos anything for you my man)))

Yes, I have artifacts on output. Yes, I have question marks (?) at the beginning of the lines, just check my previous screenshots. But colouring works.

MetaMmodern commented:
@Martinsos 
![image](https://user-images.githubusercontent.com/42899786/177801733-1c219910-4c23-4ebf-8ce9-0a1a3afc5771.png)
looks fine to me. well, except the question marks.

Martinsos commented:
> @Martinsos ![image](https://user-images.githubusercontent.com/42899786/177801733-1c219910-4c23-4ebf-8ce9-0a1a3afc5771.png) looks fine to me. well, except the question marks.

Mmmmm ok, those should be emojis but I guess that is not supported on Windows -> we can work on that in the future as a separate issue!

MetaMmodern commented:
@Martinsos are you sure? I just tested emoji output, works fine. 
![image](https://user-images.githubusercontent.com/42899786/178097432-98758d7d-8bf1-4eb3-be3f-a8b17d1d4745.png)

Martinsos commented on waspc/src/Wasp/Generator/Common.hs:
```diff
@@ -38,3 +43,36 @@ npmVersionRange =
+  _ -> (npxCmd, args)
+
+wrapCmdAndArgsInWinCmdExe :: (String, [String]) -> (String, [String])
+wrapCmdAndArgsInWinCmdExe (cmd, args) = ("cmd.exe", [unwords $ "/c" : cmd : args])
```
We should also consider escpaing the arguments here -> we are not doing any, and we might want to do some, to both make it more robust and reduce chance of injections.

Also, if we are using `cmd.exe`, from what I am getting, we might not have to provide it with `npm.cmd` but just `npm` might be enough, and it will know how to resolve that? Is that true? Investigate this.
